File: pa4/Dictionary.c
/* 
*  Connor Monson cmonson@ucsc.edu
*  pa5
*  Dictionary.c 
*  Implements Dictionary.h using Hash Tables
*/  

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"
#define MAX_LEN 180

const int tableSize=101;

// Hash Functions

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
   int sizeInBits = 8*sizeof(unsigned int);
   shift = shift & (sizeInBits-1);
   if ( shift == 0 )
      return value;
   return (value << shift) | (value >> (sizeInBits -shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) { 
   unsigned int result = 0xBAE86554;
   while (*input) { 
      result ^= *input++;
      result = rotate_left(result, 5);
   }
   return result;
}
   
// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
   return pre_hash(key)%tableSize;
}
// -------------------------------------------------------------------------------

//NodeObj
typedef struct NodeObj {
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

//Node
typedef NodeObj* Node;

//newNode() constructor for the Node type
Node newNode(char*k, char* v){
   Node M = malloc(sizeof(NodeObj));
   assert(M!=NULL);
   M->key = k;
   M->value = v;
   M->next = NULL;
   return(M);
}

//freeNode() destructor for the Node type
void freeNode(Node* fN) {
   if (fN!=NULL && *fN!=NULL) {
      free(*fN);
      *fN = NULL;
   }
}

//ListObj
typedef struct ListObj {
   Node head;
} ListObj;

//List
typedef ListObj* List;

//newList() constructor for the Node type
List newList(void) {
   List P = malloc(sizeof(ListObj));
   assert(P!=NULL);
   P->head = NULL;
   return P;
}

//DictionaryObj
typedef struct DictionaryObj {
   List table;
   int itemCount;
} DictionaryObj;

//Private helper functions-----------------------

//findKey()
//returns a reference to the NodeObj containing its respective key
//returns null if no such Node exists
Node findKey(Node H, char* targetKey) {
   while (H!=NULL) {
      if (strcmp(H->key, targetKey)==0) {
         break;
      } 
      H = H->next;
   } return H;
}

//deleteAll
void deleteAll(Node M) {
   if (M!=NULL) {
      deleteAll(M->next);
      freeNode(&M);
   }
}

//Public functions-------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void) {
   Dictionary T = malloc(sizeof(DictionaryObj));
   assert(T!=NULL);
   T->table = calloc(tableSize, sizeof(ListObj));
   T->itemCount = 0;
   return T;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* fD) {
   if (fD!=NULL && *fD!=NULL){
      if (!isEmpty(*fD)) makeEmpty(*fD);
      free(*fD);
      *fD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary T){
   if(T==NULL){
      fprintf(stderr, "DICTIONARY ERROR: calling isEmpty() on invalid Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(T->itemCount==0);
}

// size()
// returns the number of (key, value) pairs in T
// pre: none
int size(Dictionary T){
   return T->itemCount;
}

// lookup()
// returns the value v such that (k, v) is in T, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary T, char* k){
   int tableIndex;
   Node M;
   List P;
   tableIndex = hash(k);
   P = &T->table[tableIndex];
   M = findKey(P->head, k);
   if(M==NULL) {
      return NULL;
   } else {
      return M->value;
   }
}

// insert()
// inserts new (key,value) pair into T
// pre: lookup(T, k)==NULL
void insert(Dictionary T, char* k, char* v){
   Node M;
   List P;
   int tableIndex;
   tableIndex = hash(k);
   M = newNode(k,v);
   P = &T->table[tableIndex];
   if(findKey(P->head,k)!=NULL) {
      fprintf(stderr, "DICTIONARY ERROR: calling insert() on an existing key\n");
      exit(EXIT_FAILURE);
   }
   M->next = P->head;
   P->head = M;
   M = NULL;
   T->itemCount++;
}

// delete()
// deletes pair with the key k
// pre: lookup(T, k)!=NULL
void delete(Dictionary T, char* k) {
   if (T==NULL) {
      fprintf(stderr, "DICTIONARY ERROR: calling delete() on an invalid Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   Node M;
   List P;
   int tableIndex;
   tableIndex = hash(k);
   P = &T->table[tableIndex];
   if (findKey(P->head, k)==P->head) {
      M = P->head;
      P->head = P->head->next;
      M->next = NULL;
   } else {
      M = findKey(P->head,k);
      Node prev = P->head;
      Node temp = P->head->next;
      while (temp != M) {
         temp = temp->next;
         prev = prev->next;
      }
      prev->next = M->next;
      M->next = NULL;
   }
   T->itemCount--;
   freeNode(&M);
}

// makeEmpty()
// re-sets T to the empty state.
// pre: none
void makeEmpty(Dictionary T) {
   List P;
   if (T==NULL) {
      fprintf(stderr, "DICTIONARY ERROR: calling makeEmpty() on an invalid Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if (T->itemCount==0) {
      fprintf(stderr, "DICTIONARY ERROR: calling makeEmpty() on an empty Dictionary\n");
      exit(EXIT_FAILURE);
   }
   for (int i=0; i<tableSize; i++) {
      P = &T->table[i];
      deleteAll(P->head);
   }
   T->table = NULL;
   T->itemCount = 0;
}

// printDictionary()
// pre: none
// prints a text representation of T to the file pointed to by out
void printDictionary(FILE* out, Dictionary T) {
   if (T==NULL) {
      fprintf(stderr, "DICTIONARY ERROR: calling printDictionary() on an invalid Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   Node M;
   List P;
   for (int i=0; i<tableSize; i++) {
      P = &T->table[i];
      M = P->head;
      while (M!=NULL) {
         fprintf(out, "%s %s\n", M->key, M->value);
         M = M->next;
      }
   }
}

File: Lab7/Dictionary.java
// Connor Monson cmonson@ucsc.edu
// lab7 
// Dictionary.java
// Implements Dictionary ADT using BST


public class Dictionary implements DictionaryInterface {

   //private inner Node class
   private class Node {
      Pair item;
      Node left;
      Node right;

      Node(Pair x) {
         item = x;
         left = null;
         right = null;
      }
   }

   //private inner Pair class
   private class Pair {
      String key;
      String value;

      Pair(String x, String y) {
         key = x;
         value = y;
      }
   }

   //fields for Dictionary class
   private Node root;      //root is the first Node in List
   private int itemCount;   //itemCount in number of Dictionary items

   //Dictionary() constructor
   public Dictionary() {
      root = null;
      itemCount = 0;
   }

   //private helper functions----------------------------------------------------

   //findKey()
   //returns Node that has key, k, in subtree root
   //returns NULL if no such Node exists
   private Node findKey(Node R, String targetKey) {
      if (R==null || R.item.key.equals(targetKey)) {
         return R;
      }
      if (R.item.key.compareToIgnoreCase(targetKey)>0) {
         return findKey(R.left, targetKey);
      } else {
         return findKey(R.right, targetKey);
      }   
   }	 	

   //findParent()
   //pre: R!=NULL
   //returns parent of N in the subtree rooted R
   //returns NULL if N is equal to R
   Node findParent(Node N, Node R) {
      Node E = null;
      if (N!=R) {
         E=R;
         while (E.left!=N && E.right!=N) {
            if (N.item.key.compareToIgnoreCase(E.item.key)<0) {
               E = E.left;
            } else {
               E = E.right;
            }
         }
      }
      return E;
   }

   //findLeftmost()
   //returns leftmost Node in the subtree rooted R
   //returns NULL if R is NULL
   Node findLeftmost(Node R) {
      Node L = R;
      if ( L!= null ) for ( ; L.left != null; L = L.left);
         return L;
   }

   // printInOrder()
   // prints (key, value) pairs from the subtree rooted at R in ascending order
   void printInOrder(Node R) {
      if ( R != null ) {
         printInOrder(R.left);
         System.out.println(R.item.key + " " + R.item.value);
         printInOrder(R.right);
      }
   }

   // deleteAll()
   // deletes all Nodes in the subtree rooted at N.
   void deleteAll(Node N) {
      if ( N != null ) {
         deleteAll(N.left);
         deleteAll(N.right);
      }
   }

   // ADT operations ------------------------------------------------------------

   // isEmpty()
   // pre: none
   // post: returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty() {
      return (itemCount == 0);
   }

   // size()
   // pre: none
   // post: returns the number of elements in this Dictionary
   public int size() {
      return itemCount;
   }

   // lookup
   // pre: none
   // post: returns value associated key, or null reference if no such key exists
   public String lookup(String key) {
      Node N;
      N = findKey(root, key);
      return ( N == null ? null : N.item.value );
   }

   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   // post: !isEmpty(), size() is increased by one
   public void insert(String key, String value) throws DuplicateKeyException {
      Node N, A, B;
      if (findKey(root, key)!=null) {
         throw new DuplicateKeyException("Dictionary Error: insert() cannot insert duplicate keys");
      }
      N = new Node(new Pair(key, value));
      B = null;
      A = root;
      while ( A != null ) {
         B = A;
         if (A.item.key.compareToIgnoreCase(key)>0) {
            A = A.left;
         } else {
            A = A.right;
         }
      }
      if ( B == null ) {
         root = N;
      } else if (B.item.key.compareToIgnoreCase(key)>0) {
         B.left = N;
      } else {
         B.right = N;
      }
      itemCount++;
   }

   // delete()
   // deletes pair with the given key
   // pre: lookup(key)!=null
   public void delete(String key) throws KeyNotFoundException {
      Node N, E, S;
      if (findKey(root, key)==null) {
         throw new KeyNotFoundException("Dictionary Error: delete() cannot delete non-existent key");
      }
      N = findKey(root, key);
      if ( N.left == null && N.right == null ) {
         if ( N == root ) {
            root = null;
         } else {
            E = findParent(N, root);
            if ( E.right == N ) {
               E.right = null;
            } else {
               E.left = null;
            }
         }
      } else if ( N.right == null ) {
         if ( N == root ) {
            root = N.left;
         } else {
            E = findParent(N, root);
            if ( E.right == N ) {
               E.right = N.left;
            } else {
               E.left = N.left;
            }     
         }
      } else if ( N.left == null ) {
         if ( N == root ){
            root = N.right;
         } else {
            E = findParent(N, root);
            if ( E.right == N ) {
               E.right = N.right;
            } else {
               E.left = N.right;
            }
         }
      } else {  
         S = findLeftmost(N.right);
         N.item.key = S.item.key;
         N.item.value = S.item.value;
         E = findParent(S, N);
         if ( E.right == S ) {
            E.right = S.right;
         } else {
            E.left = S.right;
         }
      }
      itemCount--;
   }

   // makeEmpty
   // pre: none
   public void makeEmpty() {
      deleteAll(root);
      root = null;
      itemCount = 0;
   }

   // toString()
   // Overrides Object's toString() method
   public String toString() {
      printInOrder(root);
      return "";
   }
}